File: credit.c
#include <stdio.h>
#include <cs50.h>


long long numberofDigits(long long num);
void findCard(int x, int y, long long number);
int main(void)
{
    int sum = 0;
    long long number = get_long("Number: ");
    long long d = numberofDigits(number);
    int product[d]; //make an array called products with the same number of digits as the input
    long long newNum = number;
    for(int j=0; j<d; j++) //Create in array with all of the input numbers
    {
        product[j] = newNum%10;
        newNum /=10;//remove the last digit
    }
    for(int i =1; i<d; i+=2) //Run it one time for every digit, but skip every other digit
    {
        if((product[i]*2)>9)//If doubling make it two digits add each digit individually
        {
            int product1 = (product[i]*2)%10;
            int product2 = (product[i]*2)/10;
            sum = sum+product1+product2;
        }
        else
        {
        sum += product[i]*2; //Mulptiply every other digit by 2 and add them together
        }
    }
    int leftovers = 0;
    for (int q=0; q<d; q+=2)//Add up all the left over digits
    {
        leftovers+=product[q];
    }
    int digit = (int) sum + leftovers;//
    if(digit%10==0) //If the last digit of that = 0
    {
        findCard(product[d-1], product[d-2], number); //Find which card company it is
    }
    else //If it's not 0 it's not a valid card
    {
        printf("INVALID\n");
    }
}

void findCard(int x, int y, long long number)
{
        int d = numberofDigits(number);
        printf("number of digits: %i\n", d);
        if(x==5 && (y>=1 && y<=5) && d==16) //Mastercard
        {
            printf("MASTERCARD\n");
        }
        else if (x==3 && (y==4 || y==7) && d==15) //American Express
        {
            printf("AMEX\n");
        }
        else if(x==4 && (d==13 || d==16)) //VISA
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }

}


long long numberofDigits(long long num)
{
    long long n=0;
    do
    {
        n++; //Add one digit
        num/=10; //Take a digit off n
    } while(num!=0); //Do it until there are no digits left
    return n;
}

File: credit.py
from cs50 import get_int

def main():
    digitSum = 0
    leftovers = 0
    product = [] # Will store the needed products
    number = get_int("Number: ") # Prompts user for number
    digits = numberOfDigits(number) # Calculates number of digits
    print(f"Number of digits: {digits}") # Make sure it adds digits correctly
    tempNumber = number # Have to convert to string to get the ith element
    for i in range(digits): # Create list with the digits of the number in it
        product.append(tempNumber % 10)
        tempNumber //= 10
    for j in range(1, digits, 2): # Multiply ever other digit and add them together
        if product[j] * 2 > 9:
            product1 = product[j] * 2 % 10  
            product2 = product[j] *2 // 10
            digitSum += product1 + product2
        else:
            digitSum += product[j] * 2
    for k in range(0, digits, 2): # Add up leftover digits
        leftovers += product[k]
    sumOfAll = digitSum + leftovers
    if sumOfAll % 10 == 0:
        findCard(product[digits - 1], product[digits - 2 ], digits) # Find card company and print
    else:
        print("INVALID")


def findCard(x, y, digits):
    if x == 5 and (y >= 1 and y<= 5) and digits == 16:
        print("MASTERCARD")
    elif x == 3 and (y == 4 or y == 7) and digits ==15:
        print("AMEX")
    elif x == 4 and (digits == 13 or digits == 16):
        print("VISA")
    else:
        print("INVALID")


def numberOfDigits(number):
    tempDigits = 0
    while number != 0: # Count all the digits
        tempDigits += 1 # Add one to the number of digits
        number //= 10 # Take one digit off
    return tempDigits # Return the number of digits

main()
